android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java

```diff
@@ -205,7 +205,63 @@ public void setIsChiral(boolean isChiral) {
   }
 
   @JavascriptInterface
-  public abstract boolean dispatchKey(final int code, final int eventModifiers);
+  public boolean dispatchKey(final int code, final int eventModifiers) {
+    Handler mainLoop = new Handler(Looper.getMainLooper());
+    mainLoop.post(new Runnable() {
+      public void run() {
+        if (k == null) {
+          KMLog.LogError(TAG, "dispatchKey failed: Keyboard is null");
+          return;
+        }
+
+        if (k.subKeysWindow != null) {
+          return;
+        }
+
+        if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP &&
+            (KMTextView.activeView == null || KMTextView.activeView.getClass() != KMTextView.class)) {
+          if (KMTextView.activeView == null && KMManager.isDebugMode()) {
+            Log.w(TAG, "dispatchKey failed: activeView is null");
+          }
+          return;
+        }
+
+        InputConnection ic = (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) ?
+          KMTextView.activeView.onCreateInputConnection(new EditorInfo()) :
+          KMManager.getInputMethodService().getCurrentInputConnection();
+        if (ic == null) {
+          if (KMManager.isDebugMode()) {
+            Log.w(TAG, "insertText failed: InputConnection is null");
+          }
+          return;
+        }
+
+        k.dismissHelpBubble();
+        k.setShouldShowHelpBubble(false);
+
+        // Handle tab or enter since KMW didn't process it
+        Log.d(TAG, "dispatchKey called with code: " + code + ", eventModifiers: " + eventModifiers);
+        if (code == KMScanCodeMap.scanCodeMap[KMScanCodeMap.KEY_TAB]) {
+          KeyEvent event = new KeyEvent(0, 0, 0, KeyEvent.KEYCODE_TAB, 0, eventModifiers, 0, 0, 0);
+          if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) {
+            KMTextView textView = (KMTextView)KMTextView.activeView;
+            textView.dispatchKeyEvent(event);
+          } else {
+            ic.sendKeyEvent(event);
+          }
```

Comment:
This also feels like it should be a helper function, especially as it is repeated immediately below! c.f. `keyDownUp`

Reply:
Refactored to use keyDownUp. I included special handling for InApp with KEYCODE_TAB.

Otherwise, dispatchKeyEvent(KEYCODE_TAB) was inserting newlines in the Keyman app.